// topology-core/nodes/arrow.rect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var rect_1 = require("../../models/rect");
function leftArrowIconRect(node) {
    node.iconRect = new rect_1.Rect(0, 0, 0, 0);
}
exports.leftArrowIconRect = leftArrowIconRect;
function leftArrowTextRect(node) {
    node.textRect = new rect_1.Rect(node.rect.x + node.rect.height / 2, node.rect.y + node.rect.height / 3, node.rect.width - node.rect.height / 2, node.rect.height / 3);
    node.fullTextRect = node.textRect;
}
exports.leftArrowTextRect = leftArrowTextRect;
function rightArrowIconRect(node) {
    node.iconRect = new rect_1.Rect(0, 0, 0, 0);
}
exports.rightArrowIconRect = rightArrowIconRect;
function rightArrowTextRect(node) {
    node.textRect = new rect_1.Rect(node.rect.x, node.rect.y + node.rect.height / 3, node.rect.width - node.rect.height / 2, node.rect.height / 3);
    node.fullTextRect = node.textRect;
}
exports.rightArrowTextRect = rightArrowTextRect;
function twowayArrowIconRect(node) {
    node.iconRect = new rect_1.Rect(0, 0, 0, 0);
}
exports.twowayArrowIconRect = twowayArrowIconRect;
function twowayArrowTextRect(node) {
    node.textRect = new rect_1.Rect(node.rect.x + node.rect.height / 2, node.rect.y + node.rect.height / 3, node.rect.width - node.rect.height, node.rect.height / 3);
    node.fullTextRect = node.textRect;
}
exports.twowayArrowTextRect = twowayArrowTextRect;
