// topology-core/models/data.js
"use strict";
Object.defineProperty(exports, "__esModule", {
  value: true
});
var node_1 = require("./node");
var line_1 = require("./line");
var status_1 = require("./status");
var TopologyData = /** @class */ (function () {
  function TopologyData(json) {
    this.pens = [];
    this.lineName = 'curve';
    this.fromArrowType = '';
    this.toArrowType = 'triangleSolid';
    this.scale = 1;
    this.locked = status_1.Lock.None;
    if (json) {
      this.pens = [];
      for (var _i = 0, _a = json.pens; _i < _a.length; _i++) {
        var item = _a[_i];
        if (item.from) {
          this.pens.push(new line_1.Line(item));
        } else {
          this.pens.push(new node_1.Node(item));
        }
      }
      this.lineName = json.lineName || 'curve';
      this.fromArrowType = json.fromArrowType || '';
      this.toArrowType = json.toArrowType || 'triangleSolid';
      this.scale = json.scale || 1;
      this.locked = json.locked || status_1.Lock.None;
      this.bkImage = json.bkImage;
      this.bkColor = json.bkColor;
      this.grid = json.grid;
      if (typeof json.data === 'object') {
        this.data = JSON.parse(JSON.stringify(json.data));
      } else {
        this.data = json.data || '';
      }
    }
  }
  return TopologyData;
}());
exports.TopologyData = TopologyData;
